#pragma once

#include <mapping/mapping.h>
#include <ros/ros.h>

#include <Eigen/Core>
#include <queue>

namespace prediction {

struct Node {
  Eigen::Vector3d p, v, a;
  double t;
  double score;
  double h;
  Node* parent = nullptr;
};
typedef Node* NodePtr;
class NodeComparator {
 public:
  bool operator()(NodePtr& lhs, NodePtr& rhs) {
    return lhs->score + lhs->h > rhs->score + rhs->h;
  }
};
struct Predict {
 private:
  static constexpr int MAX_MEMORY = 1 << 22;
  // searching

  double dt;
  double pre_dur;
  double rho_a;
  double vmax;
  mapping::OccGridMap map;
  NodePtr data[MAX_MEMORY];
  int stack_top;

  inline bool isValid(const Eigen::Vector3d& p, const Eigen::Vector3d& v) const {
    return (v.norm() < vmax) && (!map.isOccupied(p));
  }

 public:
  inline Predict(ros::NodeHandle& nh) {
    nh.getParam("tracking_dur", pre_dur);
    nh.getParam("tracking_dt", dt);
    nh.getParam("prediction/rho_a", rho_a);
    nh.getParam("prediction/vmax", vmax);
    for (int i = 0; i < MAX_MEMORY; ++i) {
      data[i] = new Node;
    }
  }
  inline void setMap(const mapping::OccGridMap& _map) {
    map = _map;
    // map.inflate_last();
  }

  inline bool predict(const Eigen::Vector3d& target_p,
                      const Eigen::Vector3d& target_v,
                      std::vector<Eigen::Vector3d>& target_predcit,
                      std::vector<Eigen::Vector3d>& target_vel,
                      const double& max_time = 0.1) {
    auto score = [&](const NodePtr& ptr) -> double {
      return rho_a * ptr->a.norm();
    };
    Eigen::Vector3d end_p = target_p + target_v * pre_dur;
    auto calH = [&](const NodePtr& ptr) -> double {
      return 0.001 * (ptr->p - end_p).norm();
    };
    ros::Time t_start = ros::Time::now();
    std::priority_queue<NodePtr, std::vector<NodePtr>, NodeComparator> open_set;
    Eigen::Vector3d input(0, 0, 0);

    stack_top = 0;
    NodePtr curPtr = data[stack_top++];
    curPtr->p = target_p;
    curPtr->v = target_v;
    curPtr->a.setZero();
    curPtr->parent = nullptr;
    curPtr->score = 0;
    curPtr->h = 0;
    curPtr->t = 0;
    double dt2_2 = dt * dt / 2;
    if(!isValid(curPtr->p, curPtr->v))
    {
        std::cout << "[prediction] start node invalid, may cause failure !" << std::endl;
    }
    while (curPtr->t < pre_dur) {
      for (input.x() = -3; input.x() <= 3; input.x() += 3)
        for (input.y() = -3; input.y() <= 3; input.y() += 3) 
            {
            bool valid = true;
            // int check_sum = 5;
            // double temp_t;
            // Eigen::Vector3d temp_p;
            // for(int check_n = 1; check_n < check_sum;  check_n++)
            // {
            //     temp_t = dt * check_n / check_sum;
            //     // temp_p = curPtr->p + curPtr->v * temp_t + 0.5 * input * temp_t * temp_t;
            //     temp_p = curPtr->p + curPtr->v * temp_t + input * temp_t * temp_t / 2.0;
            //     if(map.isOccupied(temp_p))
            //         valid = false;
            // }  
            Eigen::Vector3d p = curPtr->p + curPtr->v * dt + input * dt2_2;
            Eigen::Vector3d v = curPtr->v + input * dt;
            if(!valid){
                continue;
            }
            if (!isValid(p, v)) {
                continue;
            }
            if (stack_top == MAX_MEMORY) {
                std::cout << "[prediction] out of memory!" << std::endl;
                return false;
            }
            double t_cost = (ros::Time::now() - t_start).toSec();
            if (t_cost > max_time) {
                std::cout << "[prediction] too slow!" << std::endl;
                return false;
            }
            NodePtr ptr = data[stack_top++];
            ptr->p = p;
            ptr->v = v;
            ptr->a = input;
            ptr->parent = curPtr;
            ptr->t = curPtr->t + dt;
            ptr->score = curPtr->score + score(ptr);
            ptr->h = calH(ptr);
            open_set.push(ptr);
            // std::cout << "open set push: " << state.transpose() << std::endl;
            }
      if (open_set.empty()) {
        std::cout << "[prediction] no way!" << std::endl;
        return false;
      }
      curPtr = open_set.top();
      open_set.pop();
    }
    target_predcit.clear();
    target_vel.clear();
    while (curPtr != nullptr) {
      target_predcit.push_back(curPtr->p);
      target_vel.push_back(curPtr->v);
      curPtr = curPtr->parent;
    }
    std::reverse(target_predcit.begin(), target_predcit.end());
    std::reverse(target_vel.begin(), target_vel.end());
    return true;
  }
};

}  // namespace prediction